### PR TITLE
ChartUtils: GetTicks only returns valid ticks that are part of the current domain.

### PR DESCRIPTION
**Pull request by nikolasrieble**
## Description
Description of the issue is found here. https://github.com/recharts/recharts/issues/3447#issuecomment-1465206498

The solution is simple: 
Filter all ticks that have NaN as a coordinate. 

This PR additionally adds unit tests and a few comments.

## Related Issue
https://github.com/recharts/recharts/issues/3447

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/cartesian/CartesianAxis.tsx [hunk 1]
```diff
@@ -315,10 +315,14 @@ export class CartesianAxis extends Component<Props, IState> {
     const { ticks, ...noTicksProps } = this.props;
     let finalTicks = ticks;
 
+    console.log('before', finalTicks);
+
```
Comment (nikolasrieble) [inline src/cartesian/CartesianAxis.tsx:319]:
```suggestion
```

File: src/cartesian/CartesianAxis.tsx [hunk 2]
```diff
@@ -315,10 +315,14 @@ export class CartesianAxis extends Component<Props, IState> {
...
     if (_.isFunction(ticksGenerator)) {
       finalTicks = ticks && ticks.length > 0 ? ticksGenerator(this.props) : ticksGenerator(noTicksProps);
     }
 
+    console.log('after', finalTicks);
+
```
Comment (nikolasrieble) [inline src/cartesian/CartesianAxis.tsx:325]:
```suggestion
```

File: src/util/ChartUtils.ts [hunk 3]
```diff
@@ -573,27 +573,44 @@ export const getCoordinatesOfGrid = (ticks: Array<TickItem>, min: number, max: n
...
+    realScaleType?: 'scaleBand' | 'band' | 'point' | 'linear';
+    scale?: any;
+    axisType?: AxisType;
+    ticks?: any;
+    niceTicks?: any;
+    isCategorical?: boolean;
+    categoricalDomain?: any;
+  },
+  isGrid?: boolean,
+  isAll?: boolean,
```
Comment (nikolasrieble) [inline src/util/ChartUtils.ts:588]: While I do not like `any` at all, this is a marginal improvement over the current type.

File: src/util/ChartUtils.ts [hunk 4]
```diff
@@ -573,27 +573,44 @@ export const getCoordinatesOfGrid = (ticks: Array<TickItem>, min: number, max: n
...
+): TickItem[] | null => {
   if (!axis) return null;
   const { scale } = axis;
   const { duplicateDomain, type, range } = axis;
 
   const offsetForBand = axis.realScaleType === 'scaleBand' ? scale.bandwidth() / 2 : 2;
   let offset = (isGrid || isAll) && type === 'category' && scale.bandwidth ? scale.bandwidth() / offsetForBand : 0;
 
-  offset = axis.axisType === 'angleAxis' ? mathSign(range[0] - range[1]) * 2 * offset : offset;
+  offset = axis.axisType === 'angleAxis' && range?.length >= 2 ? mathSign(range[0] - range[1]) * 2 * offset : offset;
```
Comment (nikolasrieble) [inline src/util/ChartUtils.ts:597]: Fixing an unrelated bug for when the range does not contain range[0], and range[1]

File: src/util/ChartUtils.ts [hunk 5]
```diff
@@ -573,27 +573,44 @@ export const getCoordinatesOfGrid = (ticks: Array<TickItem>, min: number, max: n
...
       return {
+        // If the scaleContent is not a number, the coordinate will be NaN.
+        // That could be the case for example with a PointScale and a string as domain.
         coordinate: scale(scaleContent) + offset,
         value: entry,
         offset,
       };
     });
+
+    return result.filter((row: TickItem) => !_.isNaN(row.coordinate));
```
Comment (nikolasrieble) [inline src/util/ChartUtils.ts:613]: Finally the fix.

File: src/util/ChartUtils.ts (hunk 3 above)
Comment (ckifer) [reply src/util/ChartUtils.ts:588]: totally

File: src/util/ChartUtils.ts (hunk 4 above)
Comment (ckifer) [reply src/util/ChartUtils.ts:597]: why `>=` and not just `=== 2`?

File: src/util/ChartUtils.ts (hunk 5 above)
Comment (ckifer) [reply src/util/ChartUtils.ts:613]: So there is no way to fix this upstream just because of the way that `PointScale` works?

**Review by ckifer — APPROVED** [review]
Niceeee!

File: src/util/ChartUtils.ts (hunk 5 above)
Comment (nikolasrieble) [reply src/util/ChartUtils.ts:613]: Upstream nothing is broken. This is about the domain being filtered by the brush - as it should - and the PointScale simply has no coordinates for items outside of the domain. 

That is how the point scale works.

Comment (nikolasrieble) [reply src/util/ChartUtils.ts:613]: To my best understanding, all works as it should - but this function.

File: src/util/ChartUtils.ts (hunk 4 above)
Comment (nikolasrieble) [reply src/util/ChartUtils.ts:597]: Fair enough, ===2 would also work, but it also is stricter, thus I am not certain it might break something.

Comment (nikolasrieble) [reply src/util/ChartUtils.ts:597]: >= on the other only prevents a bug without restricting beyond the requirement.

Comment (ckifer) [reply src/util/ChartUtils.ts:597]: yeah I was curious if there ever was going to be more than 2. Hopefully not but agreed

File: src/util/ChartUtils.ts (hunk 5 above)
Comment (ckifer) [reply src/util/ChartUtils.ts:613]: Got it. Logically in my head I think I'd want the data itself to be filtered and then everything trickle down into domain and ticks from there. Makes sense, but these are the kind of fragile pieces we have to keep an eye on and harden/test

Comment (ckifer) [reply src/util/ChartUtils.ts:613]: Also do we want to use `_.isNaN(row.coordinate)` as opposed to `Number.isNaN(row.coordinate)`? The more we use lodash the more we have dependency on it (I know its used a ton but just future thinking)

**Pull request merged**